packages/fetchai/skills/tac_control_contract/behaviours.py

```diff
@@ -0,0 +1,271 @@
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""This package contains a the behaviours."""
+
+import datetime
+import logging
+from typing import Optional, cast
+
+from aea.contracts.ethereum import Contract
+from aea.crypto.ethereum import EthereumApi
+from aea.decision_maker.messages.transaction import TransactionMessage
+from aea.helpers.search.models import Attribute, DataModel, Description
+from aea.skills.base import Behaviour
+
+from packages.fetchai.protocols.oef.message import OEFMessage
+from packages.fetchai.protocols.oef.serialization import DEFAULT_OEF, OEFSerializer
+from packages.fetchai.protocols.tac.message import TACMessage
+from packages.fetchai.protocols.tac.serialization import TACSerializer
+from packages.fetchai.skills.tac_control_contract.game import Game, Phase
+from packages.fetchai.skills.tac_control_contract.parameters import Parameters
+
+CONTROLLER_DATAMODEL = DataModel(
+    "tac",
+    [Attribute("version", str, True, "Version number of the TAC Controller Agent.")],
+)
+
+logger = logging.getLogger("aea.tac_control_skill")
+
+
+class TACBehaviour(Behaviour):
+    """This class implements the TAC control behaviour."""
+
+    def __init__(self, **kwargs):
+        """Instantiate the behaviour."""
+        super().__init__(**kwargs)
+        self._oef_msg_id = 0
+        self._registered_desc = None  # type: Optional[Description]
+
+    def setup(self) -> None:
+        """
+        Implement the setup.
+
+        :return: None
+        """
+        parameters = cast(Parameters, self.context.parameters)
+        contract = cast(Contract, self.context.contracts.erc1155)
+        ledger_api = cast(EthereumApi, self.context.ledger_apis.apis.get("ethereum"))
+
+        #  Deploy the contract if there is no address in the parameters
+        if parameters.contract_address is None:
+            contract.set_instance(ledger_api)
+            transaction_message = contract.get_deploy_transaction(
+                deployer_address=self.context.agent_address,
+                ledger_api=ledger_api,
+                skill_callback_id=self.context.skill_id,
+            )
+
+            self.context.decision_maker_message_queue.put_nowait(transaction_message)
+        else:
+            self.context.logger.info("Setting the address of the deployed contract")
+            contract.set_instance_w_address(
+                ledger_api=ledger_api,
+                contract_address=str(parameters.contract_address),
+            )
+            contract.item_ids = parameters.good_ids
```

Comment:
Note, due to the involvement of the decision maker, there is an asymmetry in the setup here. We should ensure that the calls to the decision maker during a setup phase of a skill are completed and the returns handled before the setup phase of all skills finishes. Please open an issue with respect to this.

Reply:
I cannot fully understand what issue I should open. Is this that we should handle the deployment of the contract before opening the registration and we should take the time it need to do that in consideration before we start accepting players to the game ?